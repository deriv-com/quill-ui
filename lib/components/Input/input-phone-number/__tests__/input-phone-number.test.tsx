import React from "react";
import { render, fireEvent } from "@testing-library/react";
import InputPhoneNumber, { InputPhoneNumberProps } from "../index";

const mockCountryCodes = [
    { short_code: "US", name: "USA", phone_code: "+1" },
    { short_code: "GB", name: "GB", phone_code: "+44" },
    { short_code: "IN", name: "IN", phone_code: "+91" },
];

describe("InputPhoneNumber Component", () => {
    let defaultProps: InputPhoneNumberProps;

    beforeEach(() => {
        defaultProps = {
            countryCodes: mockCountryCodes,
            onCodeChange: jest.fn(),
            onValueChange: jest.fn(),
        };
    });

    it("should match the snapshot", () => {
        const { asFragment } = render(<InputPhoneNumber {...defaultProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should call onValueChange when input value changes", () => {
        const { getByRole } = render(<InputPhoneNumber {...defaultProps} />);
        const input = getByRole("textbox");

        fireEvent.change(input, { target: { value: "1234567890" } });
        expect(defaultProps.onValueChange).toHaveBeenCalledWith("+11234567890");
    });

    it("should call onCodeChange when country code is changed", () => {
        const { getByText } = render(
            <InputPhoneNumber
                {...defaultProps}
                shortCode="US"
                codeLabel="Code"
            />,
        );
        const dropdownButton = getByText("Code");

        fireEvent.click(dropdownButton);
        const dropdownOption = getByText("GB (+44)");
        fireEvent.click(dropdownOption);

        expect(defaultProps.onCodeChange).toHaveBeenCalledWith(
            mockCountryCodes[1],
        );
    });

    it("should update phone code and input value state", () => {
        const { getByText, getByRole } = render(
            <InputPhoneNumber
                {...defaultProps}
                shortCode="US"
                codeLabel="Code"
            />,
        );
        const input = getByRole("textbox");
        const dropdownButton = getByText("Code");

        fireEvent.click(dropdownButton);
        fireEvent.click(getByText(/91/i));
        expect(defaultProps.onCodeChange).toHaveBeenCalledWith(
            mockCountryCodes[2],
        );

        fireEvent.change(input, { target: { value: "9876543210" } });
        expect(defaultProps.onValueChange).toHaveBeenCalledWith(
            "+919876543210",
        );
    });

    it("should render custom props correctly", () => {
        const { getByRole } = render(
            <InputPhoneNumber
                inputSize="sm"
                {...defaultProps}
                status="error"
                placeholder="Enter your phone number"
            />,
        );
        const input = getByRole("textbox");
        expect(input).toBeInTheDocument();
    });
});
