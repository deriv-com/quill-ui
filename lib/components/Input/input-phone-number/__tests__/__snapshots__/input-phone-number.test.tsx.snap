// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputPhoneNumber Component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="quill-phone-input__container"
  >
    <div
      class="quill-input__container"
    >
      <div
        class="quill-input__wrapper quill-input__wrapper__variant--outline quill-input__wrapper__variant--outline--neutral quill-input__wrapper__size--lg"
      >
        <div
          class="quill-phone-input-addon"
        >
          <div
            class="quill-custom-dropdown__container"
          >
            <div
              class="quill-custom-dropdown__head"
            >
              <div
                class="quill-country-code quill-country-code__variant--outline quill-country-code__size--lg"
              >
                <svg
                  class="mobile-icon"
                  height="24"
                  role="img"
                  viewBox="0 0 32 32"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18.5 7.75v.625c0 .703-.586 1.25-1.25 1.25h-2.5c-.703 0-1.25-.547-1.25-1.25V7.75h-1.25C11.547 7.75 11 8.336 11 9v15c0 .703.547 1.25 1.25 1.25h7.5c.664 0 1.25-.547 1.25-1.25V9c0-.664-.586-1.25-1.25-1.25zm-1.25 0h-2.5v.625h2.5zM9.75 9c0-1.367 1.094-2.5 2.5-2.5h7.5c1.367 0 2.5 1.133 2.5 2.5v15c0 1.406-1.133 2.5-2.5 2.5h-7.5a2.47 2.47 0 0 1-2.5-2.5z"
                  />
                </svg>
                <div>
                  <p
                    class="quill-typography quill-typography__body-text__size--md__weight--regular__decoration--default quill-typography__color--default quill-country-code__value"
                  >
                    +1
                  </p>
                </div>
                <svg
                  class="dropdown__transform"
                  data-state="close"
                  height="24"
                  role="img"
                  viewBox="0 0 32 32"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m15.336 22.164-7.5-7.5c-.39-.351-.39-.937 0-1.328.351-.352.937-.352 1.328 0L16 20.21l6.836-6.836c.352-.39.937-.39 1.328 0a.92.92 0 0 1 0 1.29l-7.539 7.5a.856.856 0 0 1-1.29 0"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="quill-input-icons__wrapper"
        >
          <div
            class="quill-input-no-label__wrapper"
          >
            <input
              class="input peer input__align--left input__size--lg"
              id="quill-input-1"
              placeholder=" "
              type="tel"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
